consensus/votecounter/vote_counter.go

```diff
@@ -0,0 +1,194 @@
+	// various thresholds. The assumption is that a single validator cannot have voting power more than f.
+	ValidatorVotingPower(types.Height, A) types.VotingPower
+
+	// Proposer returns the proposer of the current round and height.
+	Proposer(types.Height, types.Round) A
+}
+
+type VoteCounter[V types.Hashable[H], H types.Hash, A types.Addr] struct {
+	validators        Validators[A]
+	currentHeight     types.Height
+	totalVotingPower  types.VotingPower
+	faultyVotingPower types.VotingPower
+	quorumVotingPower types.VotingPower
+	roundData         roundMap[V, H, A]
+	futureMessages    map[types.Height]roundMap[V, H, A]
+}
+
+func New[V types.Hashable[H], H types.Hash, A types.Addr](validators Validators[A], height types.Height) VoteCounter[V, H, A] {
+	totalVotingPower := validators.TotalVotingPower(height)
+	return VoteCounter[V, H, A]{
+		validators:        validators,
+		currentHeight:     height,
+		totalVotingPower:  totalVotingPower,
+		faultyVotingPower: f(totalVotingPower),
+		quorumVotingPower: q(totalVotingPower),
+		roundData:         make(roundMap[V, H, A]),
+		futureMessages:    make(map[types.Height]roundMap[V, H, A]),
+	}
+}
+
+func (v *VoteCounter[V, H, A]) LoadFromMessages(messages iter.Seq2[types.Message[V, H, A], error]) error {
+	for msg, err := range messages {
+		if err != nil {
+			return err
+		}
+		switch msg := msg.(type) {
+		case *types.Proposal[V, H, A]:
+			v.AddProposal(msg)
+		case *types.Prevote[H, A]:
+			v.AddPrevote(msg)
+		case *types.Precommit[H, A]:
+			v.AddPrecommit(msg)
+		}
+	}
+	return nil
+}
+
+func (v *VoteCounter[V, H, A]) StartNewHeight() {
+	v.currentHeight++
+	v.totalVotingPower = v.validators.TotalVotingPower(v.currentHeight)
+	v.faultyVotingPower = f(v.totalVotingPower)
+	v.quorumVotingPower = q(v.totalVotingPower)
+
+	clear(v.roundData)
+	var ok bool
+	if v.roundData, ok = v.futureMessages[v.currentHeight]; !ok {
+		v.roundData = make(roundMap[V, H, A])
+	} else {
+		delete(v.futureMessages, v.currentHeight)
+	}
+}
+
+func (v *VoteCounter[V, H, A]) getRoundDataAndVotingPower(header *types.MessageHeader[A]) (*roundData[V, H, A], types.VotingPower, bool) {
+	if header.Height < v.currentHeight {
+		return nil, 0, false
+	}
+
+	votingPower := v.validators.ValidatorVotingPower(header.Height, header.Sender)
+
+	if header.Height == v.currentHeight {
+		return getOrCreateRoundData(v.roundData, header.Round), votingPower, true
+	}
+
+	futureRoundMap, ok := v.futureMessages[header.Height]
+	if !ok {
+		futureRoundMap = make(roundMap[V, H, A])
+		v.futureMessages[header.Height] = futureRoundMap
+	}
+	return getOrCreateRoundData(futureRoundMap, header.Round), votingPower, true
+}
+
+func getOrCreateRoundData[V types.Hashable[H], H types.Hash, A types.Addr](
+	roundMap map[types.Round]*roundData[V, H, A],
+	round types.Round,
+) *roundData[V, H, A] {
+	entry, ok := roundMap[round]
+	if !ok {
+		entry = utils.HeapPtr(newRoundData[V, H, A]())
+		roundMap[round] = entry
```

Comment:
Similar comment to the one earlier regarding storing a reference in the map instead of the actual value. Marking it here to analize this case as well

Reply:
Same as above, I don't think the second line is possible, as map is not addressable.